README.md

# ASCII Graph Plotter

[![GitHub License](https://img.shields.io/badge/license-MIT-green)](LICENSE.txt)&nbsp;
[![Python Version](https://img.shields.io/badge/python-3-blue)](https://www.python.org/downloads/)&nbsp;
[![GitHub release](https://img.shields.io/github/v/release/InformaticFreak/ascii-graph-plotter)](https://github.com/InformaticFreak/ascii-graph-plotter/releases/tag/2021.1)&nbsp;
[![CodeFactor](https://www.codefactor.io/repository/github/informaticfreak/ascii-graph-plotter/badge/main)](https://www.codefactor.io/repository/github/informaticfreak/ascii-graph-plotter/overview/main)&nbsp;
![visitors](https://visitor-badge.laobi.icu/badge?page_id=informaticfreak/ascii-graph-plotter)&nbsp;
[![Example: My use case](https://img.shields.io/badge/example-my%20use%20case-orange)](http://tubaf.informatic-freak.de/hoehere-mathematik-fuer-ingenieure-1.php#epsilon-delta-kriterium-der-stetigkeit)&nbsp;

Plots the graph of a function with ASCII characters.

See the change log [here](CHANGELOG.md).

Developed by InformaticFreak (c) 2021

## How to use

```
py main.py [function: str] [scale: float,float] [origin: int,int] [size: int,int] [discontinuous: bool] [precision: bool] [ascii_only: bool]
```

| Parameter | Type | Default value | Description |
|---|---|---|---|
| *function* | `str` | | Defines the function |
| *scale* | `float,float` | `0.1,10` | Scales the graph in x,y direction |
| *origin* | `int,int` | `40,20` calculated by `(size-1)/2` | Sets the origin of the plot to the list index x,y |
| *size* | `int,int` | `81,41` | Sets the size x,y in characters of the plot |
| *discontinuous* | `bool` | `false` | Specifies if the graph should be plotted discontinuously |
| *precision* | `bool` | `false` | Specifies if the graph should be plottet with thinner characters oriented at the angle at the point |
| *ascii_only* | `bool` | `false` | Specifies whether the graph should be plotted with thinner characters, depending on the slope at each point |

### Built-In Functions

- `sin`, `cos`, `tan`, `cot`
- `asin`, `acos`, `atan`, `acot`
- `degrees`, `radian`
- `log`, `sign`, `sqrt`

### Built-In Constants

- `e`, `pi`

## Examples

- Specifie the scale in x direction as 0.2 and use the default scale in y direction of 10, so write `0.2,_`
- Set the x origin to the list index 5 and use the default y origin of list index 20, so write `5,_`
- Use the default x size of the plot of 81 and specifie the y size of the plot as 32, so write `_,32`
- Use the default for all other parameter, so don't write anything or use `_` and `_,_` as placeholder

```
py main.py sin(x) 0.2,_ 5,_ _,32
```

The result plot contains the plotted graph of the function and in the first line all parameter values to reproduce it. Also, the entire plot is copied to the clipboard.

```
f(x)=sin(x) 0.2,10.0 5,15 81,32 0 0 0

     ▲                                                                           
     │                                                                           
     │                                                                           
     │                                                                           
     │                                                                           
     ┤      ███                            ███                             ███   
     │     █   █                          █   ██                         ██   █  
     │    █     █                        █      █                       █      █ 
     │   █       █                      █       █                       █       █
     │  █         █                     █        █                     █         
     ┤  █         █                    █          █                   █          
     │ █          █                   █           █                   █          
     │█            █                  █           █                   █          
     │█             █                 █            █                 █           
     │█             █                █              █               █            
┬────█────┬────┬────█────┬────┬────┬█───┬────┬────┬─█──┬────┬────┬──█─┬────┬────▸
    █│               █              █               █               █            
    █│                █             █                █             █             
    █│                █            █                  █           █              
   █ │                 █          █                   █           █              
  █  ┤                  █         █                    █         █               
  █  │                  █        █                      █       █                
 █   │                  █       █                       █       █                
█    │                   █      █                        █     █                 
     │                    █    █                          █   █                  
     ┤                     ████                            ███                   
     │                                                                           
     │                                                                           
     │                                                                           
     │                                                                           
     ┤                                                                           
     │                                                                           
```

- Set the `discontinuous` Parameter to default, so write `_`
- Now specifie the `precision` Parameter to true, so write `1`, `true` or `True`
- Also specifie the `ascii_only` Parameter to true, so write `1`, `true` or `True`

>**Note:** Use `0`, `false` or `False` to set a Parameter of type bool to false.

```
py main.py f(x)=sin(x) 0.2,10 5,14 81,32 _ _ 1
```

Now the plot contains only real ASCII characters and the character of each point depends on its slope.

```
f(x)=sin(x) 0.2,10.0 5,15 81,32 0 1 1

     A                                                                           
     |                                                                           
     |                                                                           
     |                                                                           
     |                                                                           
     +      --\                            --\                             --\   
     |     /   \                          /   -\                         -/   \  
     |    /     \                        /      |                       /      \ 
     |   /       \                      /       \                       |       -
     |  /         |                     |        \                     /         
     +  |         |                    /          |                   /          
     | /          \                   /           |                   |          
     |/            \                  |           \                   |          
     ||             |                 |            \                 /           
     ||             |                /              |               /            
+----/----+----+----\----+----+----+/---+----+----+-|--+----+----+--|-+----+---->
    /|               \              |               \               |            
    ||                |             |                \             /             
    ||                \            /                  |           /              
   / |                 \          /                   \           |              
  /  +                  |         |                    \         /               
  |  |                  |        /                      |       /                
 /   |                  \       /                       \       |                
/    |                   \      |                        \     /                 
     |                    \    /                          \   /                  
     +                     ---/                            --/                   
     |                                                                           
     |                                                                           
     |                                                                           
     |                                                                           
     +                                                                           
     |                                                                           
```


CHANGELOG.md

# CHANGE LOG

## 2021.1

* First uploaded version


src/main_oop.py
from math import e, pi, sin, cos, tan, asin, acos, atan, log, sqrt, degrees, radians
from numpy import sign
import pyperclip as pc
import re, sys

# Plot-Class
class Plot:
	def __init__(self, size, origin):
		self.size = size
		self.origin = origin
		self.grid = [ [ None for x in range(self.size[0]) ] for y in range(self.size[1]) ]
	
	# Draw function points
	def draw(self):
		pass
	
	# Print full plot
	def show(self):
		pass



# DDA (Draw Line Algorithm)
def line(xy0, xy1):
	if xy0 == (-1, -1) or xy1 == (-1, -1): return []
	
	lp = []
	dx, dy = xy1[0] - xy0[0], xy1[1] - xy0[1]
	steps = abs(dx) if abs(dx) > abs(dy) else abs(dy)
	xi, yi = dx / steps, dy / steps
	x, y = xy0[0], xy0[1]
	
	for j in range(steps):
		lp.append((round(x), round(y)))
		x += xi
		y += yi
	
	return lp

f = eval(f"lambda x: {x}")


src/main.py
from math import e, pi, sin, cos, tan, asin, acos, atan, log, sqrt, degrees, radians
from numpy import sign
import pyperclip as pc
import re, sys

#-----------------------------------------------------------------------------------------#

__project__ = "ascii-graph-plotter"
__version__ = "2021.1"
__author__ = "InformaticFreak"
__description__ = "Plots the graph of a function with ASCII characters."

#-----------------------------------------------------------------------------------------#

# Weitere mathematische Funktionen
cot = lambda x: cos(x) / sin(x)
acot = lambda x: atan(1/x) if x > 0 else atan(1/x) + pi

# DDA (Draw Line Algorithm)
def line(xy0, xy1):
	if xy0 == (-1, -1) or xy1 == (-1, -1): return []
	
	lp = []
	dx, dy = xy1[0] - xy0[0], xy1[1] - xy0[1]
	steps = abs(dx) if abs(dx) > abs(dy) else abs(dy)
	xi, yi = dx / steps, dy / steps
	x, y = xy0[0], xy0[1]
	
	for j in range(steps):
		lp.append((round(x), round(y)))
		x += xi
		y += yi
	
	return lp

# Anstieg am Punkt xy berechnen & passendes Zeichen zuordnen: █▓▒░
def draw(xy0, xy1, high_precision=True, ascii_only=False):
	if xy0 == (-1, -1) or xy1 == (-1, -1): return " "
	
	x0, y0 = xy0
	x1, y1 = xy1
	dx, dy = (x1 - x0), (y1 - y0)
	
	if high_precision:
		m = degrees(atan(dy / dx)) if dx != 0 else degrees(atan(dy / (dx + 1e-10)))
		if -90 <= m <= -60: char = "|"
		elif -60 <= m <= -30: char = "/"
		elif -30 <= m <= 30: char = "-"
		elif 30 <= m <= 60: char = "\\"
		elif 60 <= m <= 90: char = "|"
	
	else:
		char = "#" if ascii_only else "█"
	
	return char

#-----------------------------------------------------------------------------------------#

# main.py [function: str] [scale: float,float] [origin: int,int] [size: int,int] [discontinuous: bool] [precision: bool] [ascii_only: bool]

# Feldgröße definieren, Nullpunkt position setzen & Präzission festlegen
schemes = [r"\d+,\d+", r"_,\d+", r"_,_", r"\d+,_", r"(((\d+)?\.\d+)|(\d+\.?)),(((\d+)?\.\d+)|(\d+\.?))", r"_,(((\d+)?\.\d+)|(\d+\.?))", r"_,_", r"(((\d+)?\.\d+)|(\d+\.?)),_", r"1|(T|t)rue", r"_|0|(F|f)alse"]

# Skalierung
scale_default = (0.1, 10)
scale = scale_default
try:
	if   re.search(schemes[4], sys.argv[2]): scale = (float(sys.argv[2].split(",")[0]), float(sys.argv[2].split(",")[1]))
	elif re.search(schemes[5], sys.argv[2]): scale = (scale_default[0], float(sys.argv[2].split(",")[1]))
	elif re.search(schemes[6], sys.argv[2]): scale = scale_default
	elif re.search(schemes[7], sys.argv[2]): scale = (float(sys.argv[2].split(",")[0]), scale_default[1])
except: scale = scale_default

# Feldgröße
size_default = (81, 41)
size = size_default
try:
	if re.search(schemes[0], sys.argv[4]): size = (int(sys.argv[4].split(",")[0]), int(sys.argv[4].split(",")[1]))
	elif re.search(schemes[1], sys.argv[4]): size = (size_default[0], int(sys.argv[4].split(",")[1]))
	elif re.search(schemes[2], sys.argv[4]): size = size_default
	elif re.search(schemes[3], sys.argv[4]): size = (int(sys.argv[4].split(",")[0]), size_default[1])
except: size = size_default

# Nullpunkt
origin_default = (int((size[0] - 1) / 2), int((size[1] - 1) / 2))
origin = origin_default
try:
	if re.search(schemes[0], sys.argv[3]): origin = (int(sys.argv[3].split(",")[0]), int(sys.argv[3].split(",")[1]))
	elif re.search(schemes[1], sys.argv[3]): origin = (origin_default[0], int(sys.argv[3].split(",")[1]))
	elif re.search(schemes[2], sys.argv[3]): origin = origin_default
	elif re.search(schemes[3], sys.argv[3]): origin = (int(sys.argv[3].split(",")[0]), origin_default[1])
except: origin = origin_default

# Unstetige Darstellung
discontinuous_default = False
discontinuous = discontinuous_default
try:
	if re.search(schemes[4], sys.argv[5]): discontinuous = True
	elif re.search(schemes[5], sys.argv[5]): discontinuous = discontinuous_default
except: discontinuous = discontinuous_default

# Präzision
precise_default = False
precise = precise_default
try:
	if re.search(schemes[8], sys.argv[6]): precise = True
	elif re.search(schemes[9], sys.argv[6]): precise = precise_default
except: precise = precise_default

# ASCII Darstellung
ascii_only_default = False
ascii_only = ascii_only_default
try:
	if re.search(schemes[8], sys.argv[7]): ascii_only = True
	elif re.search(schemes[9], sys.argv[7]): ascii_only = ascii_only_default
except: ascii_only = ascii_only_default

#-----------------------------------------------------------------------------------------#

# Feld initialisieren
grid = [ [ None for x in range(size[0]) ] for y in range(size[1]) ]

# Achsen in Feld markieren
chars = "-|++A>+" if ascii_only else "─│┬┤▲▸┼"
for y in range(size[1]):
	grid[y][origin[0]] = chars[3] if y % 5 == 0 else chars[1]
grid[0][origin[0]] = chars[4]

for x in range(size[0]):
	grid[origin[1]][x] = chars[2] if x % 5 == 0 else chars[0]
grid[origin[1]][size[0]-1] = chars[5]

grid[origin[1]][origin[0]] = chars[6]	

# Funktion definieren
try: f = eval(f"lambda x: {sys.argv[1][5:] if sys.argv[1].startswith('f(x)=') else sys.argv[1]}")
except IndexError: raise Exception("No function given")

# Funktionswerte ausrechnen und Position durch Nullpunkt Verschiebung ermitteln
xy = []
for x in range(size[0]):
	try: y = (-1) * round((f((x - origin[0]) * scale[0]) * scale[1]) - origin[1])
	except: y = -1
	if 0 < y < size[1]: xy.append((x, y))
	else: xy.append((-1, -1))

# Funktion approximieren (nicht unstetige Darstellung)
if not discontinuous:
	for i, c in enumerate(xy[:-1]): xy.extend(line(c, xy[i+1]))

# Funktion mit Anstieg in Feld markieren
for i, c in enumerate(xy[:-1]): grid[c[1]][c[0]] = draw(c, xy[i+1], precise, ascii_only)

# Feld in string konvertieren
output = [f"\nf(x)={sys.argv[1][5:] if sys.argv[1].startswith('f(x)=') else sys.argv[1]} {scale[0]},{scale[1]} {origin[0]},{origin[1]} {size[0]},{size[1]} {'1' if discontinuous else '0'} {'1' if precise else '0'} {'1' if ascii_only else '0'}\n\n"]
for y in grid:
	for x in y:
		if x == None: output.append(" ")
		else: output.append(str(x))
	output.append("\n")
output_str = "".join(output)

# Feld plotten & in Zwischenablage kopieren
print(output_str)
pc.copy(output_str)
